### PR TITLE
fix(DB/SAI): Garm's Bane

### LINKED ISSUE #2237: NPC's in Garm's Bane doesn't move [$15]
##### SMALL DESCRIPTION:
The NPC's doesn't move (mini Event?)

##### EXPECTED BLIZZLIKE BEHAVIOUR:
The NPCs should run towards K3

##### CURRENT BEHAVIOUR:
https://www.wowhead.com/npc=29618/snowblind-follower
https://www.wowhead.com/npc=29619/garm-invader
Video: https://www.wowhead.com/quest=12820/a-delicate-touch#videos:id=75355

##### STEPS TO REPRODUCE THE PROBLEM:
1. Fly to Garm's Bane
2. Look at the NPC's (Snowblind Follower and Garm Invader)

##### EXTRA NOTES:
idk

##### BRANCH(ES):
master

##### AC HASH/COMMIT:
0264d952e78ee0eba4d8f4e1b4bff8d8aad3b9ea

##### OPERATING SYSTEM:
Windows Server 2016

##### MODULES:
1v1 arena
better item reloading
bg reward
CF BG
CF Instances/Raids
invdividual XP
NPC CodeBox
NPC Gambler
PvPTitles
Teleporter
Weekend XP

##### OTHER CUSTOMIZATIONS:
none

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/79339934-npc-s-in-garm-s-bane-doesn-t-move?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Stoabrogga**
##### CHANGES PROPOSED:
- The Garm Invaders and Snowblind Followers will now run towards K3.
- The Garm Invaders are set active on respawn so that the NPCs also run when the player is out of reach. I ensured that the event stops if no player is near anymore by timing the kills / respawns appropriately.
- A few improvements concerning the "K3 Perimeter Turret" and the "Improved Land Mine".

##### ISSUES ADDRESSED:
Closes #2237

##### TESTS PERFORMED:
tested successfully in-game

##### HOW TO TEST THE CHANGES:
- preparation:
```
.q rem 12820
.q a 12820
.go 6299.75 -1156.48 424.188 571
```
- proceed with the quest

##### KNOWN ISSUES AND TODO LIST:
none

##### Target branch(es):
- [x] Master
 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1571086874024573263.sql [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+(29475,0,2,0,5,0,100,0,0,0,0,29618,0,33,29618,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Improved Land Mine - On Kill - Kill Credit'),
+(29475,0,3,0,5,0,100,0,0,0,0,29619,0,33,29618,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Improved Land Mine - On Kill - Kill Credit');
+
+-- Garm Invader: Set active after respawn; start random waypoint movement after 1-60 seconds;
+-- disable auto attack / combat movement while following the waypoint path
+UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 29619;
+DELETE FROM `smart_scripts` WHERE `source_type` = 0 AND `entryorguid` = 29619;
+DELETE FROM `smart_scripts` WHERE `source_type` = 9 AND `entryorguid` BETWEEN 2961900 AND 2961904;
+INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`)
+VALUES
```
Comment (Helias) [inline data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: put "VALUES" at the end of the columns list

File: data/sql/updates/pending_db_world/rev_1571086874024573263.sql [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+(29619,0,0,1,11,0,100,0,0,0,0,0,0,48,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - On Respawn - Set Active'),
+(29619,0,1,2,61,0,100,0,0,0,0,0,0,20,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - Linked - Enable Auto Attack'),
+(29619,0,2,0,61,0,100,0,0,0,0,0,0,21,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - Linked - Enable Combat Movement'),
+(29619,0,3,4,1,0,100,0,1000,60000,0,0,0,20,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - OOC - Disable Auto Attack'),
+(29619,0,4,5,61,0,100,0,0,0,0,0,0,21,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - Linked - Disable Combat Movement'),
+(29619,0,5,0,61,0,100,0,0,0,0,0,0,88,2961900,2961904,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - Linked - Call Random Action List'),
+(29619,0,6,0,58,0,100,0,0,0,0,0,0,41,1000,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - On Waypoint Path Ended - Despawn After 1 Second'),
+(29619,0,7,0,6,0,100,0,0,0,0,0,0,41,5000,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Garm Invader - On Death - Despawn After 5 Seconds'),
+
```
Comment (Helias) [inline data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: I think we can avoid this empty line, fromt he IDs a coder can guess that the smart_scripts change, do you agree?

File: data/sql/updates/pending_db_world/rev_1571086874024573263.sql (hunk 1 above)
Comment (Stoabrogga) [reply data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: Why?

Comment (Helias) [reply data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: Code style and guidelines on AC wiki http://www.azerothcore.org/wiki/Dealing-with-SQL-files
Indeed I usually see in that style on web, is a sort of standard (or standard de facto).

File: data/sql/updates/pending_db_world/rev_1571086874024573263.sql (hunk 2 above)
Comment (Stoabrogga) [reply data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: It's easier to read in my opinion by separating the script block.

Comment (Helias) [reply data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: :ok_hand:

File: data/sql/updates/pending_db_world/rev_1571086874024573263.sql (hunk 1 above)
Comment (Stoabrogga) [reply data/sql/updates/pending_db_world/rev_1571086874024573263.sql]: Just to make this clear: I did all the structuring in the SQL script on purpose to improve readability. I also put the "VALUES" in a separate line to act as a separator to the blocks. Those scripts are hard enough to read already, no need to make it more difficult.

**Comment by Knindzagxg** [comment]
Perfect one mate.
Everything works as supposed now -> Waited few times to test trigger as well and no issues at all.

**Comment by Andrei-UPB** [comment]
I've been away for a while thus didn't keep up with the awesome work you guys do.

This improvement breaks the functionality of  quest 14090 ( https://www.wowhead.com/quest=14090/gormok-wants-his-snobolds )  because it deletes the Snowblind Followers SAI entries that pertain to it, I rewrote them at the top of the SAI Script and tested it through, the quest credit is given once again. If you add the quest functionalities for Snowblind Followers after the script, for some reason they do not give credit as it should.

```

**Comment by FrancescoBorzi** [comment]
@Andrei-UPB can you please [open a PR](http://www.azerothcore.org/wiki/How-to-create-a-PR) with that?

**Comment by Stoabrogga** [comment]
@Andrei-UPB 
Yes, you're right. Sorry, but I knew that those entries were there, but played around with the SAI so long until those parts flew out 🤦‍♂. But it suffices to add the missing parts to the end of the script, I think you just forgot to update the links properly. The following script works well for me:
```sql
DELETE FROM `smart_scripts` WHERE `entryorguid` = 29618 AND `id` IN (7,8,9);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`)
VALUES
(29618,0,7,8,8,0,100,0,66474,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Snowblind Follower - On Spellhit ''Throw Net'' - Say Line 0'),
(29618,0,8,9,61,0,100,0,0,0,0,0,0,33,34899,0,0,0,0,0,7,0,0,0,0,0,0,0,0,'Snowblind Follower - On Spellhit ''Throw Net'' - Quest Credit ''Gormok Wants His Snobolds'''),
(29618,0,9,0,61,0,100,0,0,0,0,0,0,41,5000,0,0,0,0,0,1,0,0,0,0,0,0,0,0,'Snowblind Follower - On Spellhit ''Throw Net'' - Despawn In 5000 ms');
```

**Comment by Andrei-UPB** [comment]
Okay, will open a PR as soon as I am able to.

I managed to add them at the end as well, the links were fine it was just the invoker behaving in a weird way.

It is definitely working right now.

Thanks once again for the life-like improvement.

**Pull request merged**